core/stateless/witness.go

```diff
@@ -49,24 +47,23 @@ type Witness struct {
 }
 
 // NewWitness creates an empty witness ready for population.
-func NewWitness(chain HeaderReader, block *types.Block) (*Witness, error) {
-	// Zero out the result fields to avoid accidentally sending them to the verifier
-	header := block.Header()
-	header.Root = common.Hash{}
-	header.ReceiptHash = common.Hash{}
-
-	// Retrieve the parent header, which will *always* be included to act as a
-	// trustless pre-root hash container
-	parent := chain.GetHeader(block.ParentHash(), block.NumberU64()-1)
-	if parent == nil {
-		return nil, errors.New("failed to retrieve parent header")
+func NewWitness(context *types.Header, chain HeaderReader) (*Witness, error) {
+	// When building witnesses, retrieve the parent header, which will *always*
+	// be included to act as a trustless pre-root hash container
+	var headers []*types.Header
+	if chain != nil {
+		parent := chain.GetHeader(context.ParentHash, context.Number.Uint64()-1)
+		if parent == nil {
+			return nil, errors.New("failed to retrieve parent header")
+		}
+		headers = append(headers, parent)
 	}
 	// Create the wtness with a reconstructed gutted out block
 	return &Witness{
-		Block:   types.NewBlockWithHeader(header).WithBody(*block.Body()),
+		context: context,
+		Headers: headers,
 		Codes:   make(map[string]struct{}),
 		State:   make(map[string]struct{}),
```

Comment:
Great work.

Regarding the State in the block witness, does the State only include the leaf nodes (account state) involved the execution of tx in the block, or does it also include the sibling nodes of the branch nodes (used for MPT proof)?

Reply:
Should ideally include the entire partial trie?

Reply:
according https://gist.github.com/karalabe/47c906f0ab4fdc5b8b791b74f084e5f9
> Executing a single block only needs a couple hundred accounts and storage slots to be available, and verifying the final state roots only requires the proofs for those state items.

> The witness needs to contain all the data needed to execute a block (doh). This means the the accessed account/storage trie nodes (read/written slots or siblings during inserts/deletes), accessed bytecodes (via contracts ran or CODESIZE/BYTECODE) and accessed blockhashes (via BLOCKHASH). 

So yes, including the siblings to provide the entire partial trie